rewrite-xml/src/main/java/org/openrewrite/xml/ChangeTagAttribute.java

```diff
@@ -75,25 +83,31 @@ public Xml visitTag(Xml.Tag tag, ExecutionContext ctx) {
             }
 
             public Xml.Attribute visitChosenElementAttribute(Xml.Attribute attribute) {
-                if(!attribute.getKeyAsString().equals(attributeName)) {
+                if (!attribute.getKeyAsString().equals(attributeName)) {
                     return attribute;
                 }
-                if(oldValue != null && !attribute.getValueAsString().startsWith(oldValue)) {
-                    return attribute;
+
+                String stringValue = attribute.getValueAsString();
+                if (oldValue != null) {
+                    if (Boolean.TRUE.equals(regex) && !Pattern.matches(oldValue, stringValue)) {
+                        return attribute;
+                    } else if ((regex == null || Boolean.FALSE.equals(regex)) && !stringValue.startsWith(oldValue)) {
+                        return attribute;
+                    }
                 }
 
                 if (newValue == null) {
                     //noinspection DataFlowIssue
                     return null;
                 }
 
-                String changedValue = (oldValue != null) ? attribute.getValueAsString().replace(oldValue, newValue): newValue;
+                String changedValue = (oldValue != null) ? stringValue.replaceAll(oldValue, newValue) : newValue;
```

Comment:
I wonder if we should have separate the handling here between regex and non-regex code paths. `replace(a,b)` did exact replacements, whereas `replaceAll(a,b)` introduces regex matching. Seeing how `oldValue` can contain any string, that might then change the behavior around special characters like dots.

Reply:
You're right, I just remembered https://github.com/openrewrite/rewrite-logging-frameworks/pull/120. 😅